src/aleph_vrf/coordinator/vrf.py

```diff
@@ -292,3 +314,93 @@ async def publish_data(
         )
 
     return message.item_hash
+
+
+async def get_existing_vrf_message(
+    aleph_client: AuthenticatedAlephClient,
+    request_id: str,
+) -> Optional[PostMessage]:
+    channel = f"vrf_{request_id}"
+    ref = f"vrf_{request_id}"
+
+    logger.debug(
+        f"Getting VRF messages on {aleph_client.api_server} from request id {request_id}"
+    )
+
+    messages, status = await aleph_client.get_messages(
+        message_type=MessageType.post,
+        channels=[channel],
+        refs=[ref],
+    )
+
+    if messages:
+        if len(messages) > 1:
+            logger.warning(f"Multiple VRF messages found for request id {request_id}")
+        return messages[0]
+    else:
+        logger.debug(f"Existing VRF message for request id {request_id} not found")
+        return None
+
+
+async def get_existing_message(
+    aleph_client: AuthenticatedAlephClient,
+    item_hash: ItemHash,
+) -> Optional[PostMessage]:
+    logger.debug(
+        f"Getting VRF message on {aleph_client.api_server} for item_hash {item_hash}"
+    )
+
+    message, status = await aleph_client.get_message(
+        item_hash=item_hash,
+    )
+
+    if not message:
+        logger.debug(f"Existing VRF message for item_hash {item_hash} not found")
+        raise AlephNetworkError(
+            f"Message could not be read for item_hash {message.item_hash}"
+        )
+
+    return message
+
+
+async def check_message_integrity(
+    aleph_client: AuthenticatedAlephClient, message: PublishedVRFResponse
+):
+    logger.debug(
+        f"Checking VRF response message on {aleph_client.api_server} for item_hash {message.message_hash}"
+    )
+
+    for executor in message.executors:
+        generation_message = await get_existing_message(
+            aleph_client, executor.generation_message_hash
+        )
+        loaded_generation_message = PublishedVRFRandomNumberHash.from_published_message(
+            generation_message
+        )
+        publish_message = await get_existing_message(
+            aleph_client, executor.publication_message_hash
+        )
+        loaded_publish_message = PublishedVRFRandomNumber.from_published_message(
+            publish_message
+        )
+
+        if (
+            loaded_generation_message.random_number_hash
+            != loaded_publish_message.random_number_hash
+        ):
+            raise PublishedHashesDoNotMatch(
+                executor=executor,
+                generation_hash=loaded_generation_message.random_number_hash,
+                publication_hash=loaded_publish_message.random_number_hash,
+            )
+
+        if not verify(
+            HexBytes(loaded_publish_message.random_number),
+            loaded_generation_message.nonce,
+            loaded_generation_message.random_number_hash,
+        ):
+            raise PublishedHashValidationFailed(
```

Comment:
Do we really wanna raise an error if a single executor committed a crime? The XOR RNG scheme allows true randomness, as long as at least ONE of N executors contributed true randomness.

Reply:
Yes, that is what has been specified to avoid cases where "non-corrupted" nodes would be DOS-ed in order to only keep "corrupted" ones online and to avoid race conditions.

Reply:
Exactly, this was the agreement that we had on the specifications.